packages/utils/telemetry-utils/src/batchedTelemetryHelper.ts

```diff
@@ -0,0 +1,73 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import type {
+	ITelemetryGenericEventExt,
+	ITelemetryLoggerExt,
+	ITelemetryPerformanceEventExt,
+} from "./telemetryTypes.js";
+
+/**
+ * Expected type of the custom data passed into the logger.
+ */
+type ICustomDataMap = Record<string, number>;
+
+/**
+ * Helper class that accumulates custom data and sends it to the logger every time a specified number (i.e., `threshold`) of logs is reached.
+ */
+export class BatchedTelemetryHelper {
+	disposed: boolean = false;
+
+	// Stores value of the custom data passed into the logger.
+	private readonly customDataMap: Map<string, number> = new Map<string, number>();
+	// Counter to keep track of the number of times the log function is called.
+	private counter = 0;
+
+	/**
+	 * @param eventBase -
+	 * Custom properties to include in the telemetry performance event when it is written.
+	 * @param logger -
+	 * The logger to use to write the telemetry performance event.
+	 * @param threshold -
+	 * The number of logs to accumulate before sending the data to the logger.
+	 */
+	public constructor(
+		private readonly eventBase: ITelemetryGenericEventExt,
+		private readonly logger: ITelemetryLoggerExt,
+		private readonly threshold: number,
+	) {}
+
+	/**
+	 * Accumulates the custom data and sends it to the logger every time the number of logs reaches the threshold by calling `sendData()`.
+	 *
+	 * @param customData -
+	 * A record storing the custom data to be logged.
+	 */
+	public log(customData: ICustomDataMap): void {
```

Comment:
This class is keeping a single counter for all the metrics (as opposed to a counter per metric). I think that's the right choice, but it means this class will only produce correct output if the exact same set of metrics gets logged every time: if I call `log({ foo: 2 })` then `log({ bar: 10 })`, this class will compute averages 1 and 5 for foo and bar respectively.

I think it would be good to make that clearer in the contract by making `ICustomDataMap` and `BatchedTelemetryHelper ` generic over the set of metric keys.

Reply:
Once you do that, it would probably make sense to make `this.customDataMap` an `ICustomDataMap<TMetrics>` since it should contain exactly that set of values.

Reply:
Also we want to take in a read-only version of that type here: `Readonly<ICustomDataMap<TMetrics>>`